src/StatusesHandler.php

```diff
@@ -0,0 +1,270 @@
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $statusesLabels;
+
+    /**
+     * Option key name to save screen statuses config.
+     *
+     * @since [*next-version*]
+     *
+     * @var string
+     */
+    protected $screenOptionsKey;
+
+    /**
+     * Endpoint for saving statuses.
+     *
+     * @since [*next-version*]
+     *
+     * @var string
+     */
+    protected $statusesEndpoint;
+
+    /**
+     * StatusesHandler constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param Traversable           $statuses         List of statuses key in application.
+     * @param Traversable           $hiddenStatuses   List of hidden statuses in application.
+     * @param Traversable           $statusesLabels   Map of known statuses labels.
+     * @param string                $screenOptionsKey Option key name to save screen statuses config.
+     * @param string                $statusesEndpoint Endpoint for saving statuses.
+     * @param EventManagerInterface $eventManager     The event manager.
+     * @param EventFactoryInterface $eventFactory     The event factory.
+     */
+    public function __construct(
+        Traversable $statuses,
+        Traversable $hiddenStatuses,
+        Traversable $statusesLabels,
+        $screenOptionsKey,
+        $statusesEndpoint,
+        $eventManager,
+        $eventFactory
+    ) {
+        $this->statuses       = $this->_normalizeArray($statuses);
+        $this->hiddenStatuses = $this->_normalizeArray($hiddenStatuses);
+        $this->statusesLabels = $this->_normalizeArray($statusesLabels);
+
+        $this->screenOptionsKey = $screenOptionsKey;
+        $this->statusesEndpoint = $statusesEndpoint;
+
+        $this->_setEventManager($eventManager);
+        $this->_setEventFactory($eventFactory);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function __invoke()
+    {
+        /* @var $event EventInterface */
+        $event = func_get_arg(0);
+
+        if (!($event instanceof EventInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not an event instance'), null, null, $event
+            );
+        }
+
+        $event->setParams(array_merge(
+            $event->getParams(),
+            $this->_getStatusesParams()
+        ));
+    }
+
+    /**
+     * Attach events for status handling.
+     *
+     * @since [*next-version*]
+     */
+    public function run()
+    {
+        $this->_attach('eddbk_bookings_visible_statuses', function ($event) {
+            $event->setParams([
+                'statuses' => $this->_getVisibleStatuses($this->statuses, $this->hiddenStatuses),
+            ]);
+        });
+
+        /*
+         * Attach status endpoint.
+         */
+        $this->_attach('wp_ajax_set_' . $this->screenOptionsKey, function () {
+            $data = json_decode(file_get_contents('php://input'), true);
```

Comment:
I could never understand the reason behind reading from stream here. Always got the feeling that it's some acrobatic way of avoiding the use of superglobals.

Reply:
Yes, you are right. It was temp solution. I propose to create separate issue for this and solve it in another PR.